### PR TITLE
Initial acceptance and unit testing docs

### PR DESCRIPTION
**Pull request by catsby**
This is the initial chunk of acceptance and unit test documentation. Best practices / patterns will come at a latter date.

### CODE REVIEW

File: content/source/docs/extend/testing/unit-testing.html.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+The procedure for writing unit tests for Terraform follows the same setup and
+conventions of writing any Go unit tests. We recommend naming tests to follow
+the same convention as our acceptance tests, `Test<Provider>_<Test Name>`. For more
+information on Go tests, see the [official Golang docs on testing][0]. 
+
+
+Below is an example unit test used in flattening AWS security group rules,
+demonstrating a typical `flattener` type method that's commonly used to convert
+structures returned from APIs into data structures used by Terraform in saving
+to state. The blow example is truncated, but you can see the full test in the
```
Comment (appilon) [inline content/source/docs/extend/testing/unit-testing.html.md]: below

**Review by appilon — CHANGES_REQUESTED** [review]

File: content/source/docs/extend/testing/unit-testing.html.md (hunk 1 above)
Comment (catsby) [reply content/source/docs/extend/testing/unit-testing.html.md]: ?

Comment (appilon) [reply content/source/docs/extend/testing/unit-testing.html.md]: Sorry that can be interpreted in the wrong context. Typo on line "The blow example".

Comment (catsby) [reply content/source/docs/extend/testing/unit-testing.html.md]: ahhhh, yes, that's a typo. Line 24 has Below as well, so I was confused 😄 

Fixed in 41545bb93

File: content/source/docs/extend/testing/acceptance-tests/index.html.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+~> Note: When developing or testing Terraform plugins, we highly recommend
+running acceptance tests with an account dedicated to testing. This ensures no
+infrastructure is created or destroyed in error during development or validation
+of any Provider Resources in any environment that cannot be completely and
+safely destroyed.
+
+## Next Steps
+
+Terraform relies heavily on acceptance tests to ensure we keep our promise of
+helping users  safely and predictably create, change, and improve
```
Comment (appilon) [inline content/source/docs/extend/testing/acceptance-tests/index.html.md]: extra space in between "users" and "safely"

File: content/source/docs/extend/testing/acceptance-tests/testcase.html.md [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+```
+
+
+### CheckDestroy 
+
+**Type:** [TestCheckFunc](https://github.com/hashicorp/terraform/blob/0cc9e050ecd4a46ba6448758c2edc0b29bef5695/helper/resource/testing.go#L182-L186)  
+**Default:** `nil`  
+**Required:** no  
+
+**CheckDestroy** is called after the all test steps have been ran, and Terraform
```
Comment (appilon) [inline content/source/docs/extend/testing/acceptance-tests/testcase.html.md]: swap "the" and "and"

File: content/source/docs/extend/testing/acceptance-tests/teststep.html.md [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+      testAccCheckExampleResourceExists("example_widget.foo", &widgetBefore), // if testAccCheckExampleResourceExists fails to find the resource, the following TestCheckResourceAttr is still ran, with any errors aggregated
+      resource.TestCheckResourceAttr("example_widget.foo", "active", "true"),
+    ),
+  },
+},
+```
+
+## Builtin check functions
+
+Terraform has several TestCheckFunc functions builtin for developers to use for
```
Comment (appilon) [inline content/source/docs/extend/testing/acceptance-tests/teststep.html.md]: Instinctively I think "builtin" should be "built in" in this sentence but I may be wrong.

File: content/source/docs/extend/testing/acceptance-tests/teststep.html.md [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+individually, or with other `TestCheckFunc` functions with one of the above
+Aggregate functions. 
+
+It's common to write custom `TestCheckFunc` functions to validate resources were
+created correctly by using SDKs directly to verify identity and properties of
+resources. These functions can retrieve information by SDKs and provide the
+results to other `TestCheckFunc` methods. The below example uses
+`ComposeTestCheckFunc` to group a set of `TestCheckFunc` functions together. The
+first function `testAccCheckExampleWidgetExists` uses the `Example` service SDK
+directly, and queries it for the ID of the widget we have in state. After found,
```
Comment (appilon) [inline content/source/docs/extend/testing/acceptance-tests/teststep.html.md]: "After found" feels awkward to me, maybe try "Once found". Not a dealbreaker obviously it just stood out a little.

File: content/source/docs/extend/testing/acceptance-tests/teststep.html.md [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+it stores the results into the `widget` struct declared at the begining of the
+test function. The next check function `testAccCheckExampleWidgetAttributes`
+recieves the updated `widget` and check's it's attributes. The final check
```
Comment (appilon) [inline content/source/docs/extend/testing/acceptance-tests/teststep.html.md]: "checks its"

File: content/source/docs/extend/testing/index.html.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  Extending Terraform is a section for content dedicated to developing Plugins
+  to extend Terraform's core offering.
+---
+
+# Testing Terraform Plugins
+
+Here we cover information needed to write successful tests for Terraform
+Plugins. Tests are a vital part of the Terraform ecosystem, verifying we can
+deliver on our mission to safely and predictably create, change, and improve
+infrastructure. Documentation for Terraform tests is broken into categories
```
Comment (appilon) [inline content/source/docs/extend/testing/index.html.md]: "tests are" instead of "tests is"

**Review by appilon — CHANGES_REQUESTED** [review]
Super minor grammar/mistakes (some even up for debate but I thought I would still make the suggestions). Awesome stuff! So much new content!

**Comment by catsby** [comment]
Thanks @appilon ! I had to rebase this branch, and then I made the fixes you suggested. Please take another look. Also see we've added a best-practices section with your detecting drift guide

**Review by appilon — APPROVED** [review]
looking great! 👍

**Comment by catsby** [comment]
Merging! Any failing tests here are related to submodules, et. al

**Pull request merged**